### PR TITLE
improve readability of CanvasGraphics's and SVGGraphics's constructPath

### PR DESCRIPTION
**Pull request by MeiKatz**
- extract sub-programs into own functions:
  - `rectangle(x, y, width, height)`
  - `moveTo(x, y)`
  - `lineTo(x, y)`
  - `curveTo(cp1x, cp1y, cp2x, cp2y, x, y)`
  - `curveTo2(cp2x, cp2y, x, y)`
  - `curveTo3(cp1x, cp1y, x, y)`
- align CanvasGraphics's and SVGGraphics's constructPath definition
  as much as possible

## Reason
`constructPath` has now improved readability and an user searching for information about the api and the way how `PDF.js` works can now easier see the parallels between the implementation for Canvas and for SVGs.

### CODE REVIEW

**Comment by Snuffleupagus** [comment]
Won't adding a whole bunch of (strictly unnecessary) methods add an unnecessary level of indirection and quite possibly even slow down rendering by requiring *many* more function calls then the current code?

File: src/display/svg.js [hunk 1]
```diff
@@ -1001,6 +1005,90 @@ SVGGraphics = (function SVGGraphicsClosure() {
       current.setCurrentPoint(x, y);
     },
 
+    rectangle: function SVGGraphics_rectangle(x, y, width, height) {
+      var current = this.current;
+
+      if (current.path) {
```
Comment (MeiKatz) [inline src/display/svg.js]: Personally not happy about the checks in every function. Maybe we could create `current.path` at the start and ensure that it exists.

**Comment by MeiKatz** [comment]
@Snuffleupagus Yes, they're *strictly* unnecessary but they help getting what the code does. Also the methods are defined as single ops and maybe they may occur standing alone (not sure about this because I could not find any documentation about this behaviour).
I don't think that it will slow down rendering speed that much. But: my main point is that you can see with one look what the functions do. That is not the case in the current implementation IMHO.

**Comment by Snuffleupagus** [comment]
*It can often, in general, be illuminating to look at the history of a particular piece of code.*

In this case, you're essentially proposing to revert part of PR #4683, whose purpose was to improve the rendering performance of in particular large/complex PDF files. In that PR, readability was thus purposely traded for *perhaps* slightly less easily readable code.

> ... maybe they may occur standing alone (not sure about this because I could not find any documentation about this behaviour).

I don't think that can happen, given the changes made in PR #4683.

> I don't think that it will slow down rendering speed that much.

Rather than guessing/assuming, please actually test the impact of your proposed changes!
Most likely, you'll have a slightly easier time convincing a reviewer to accept your patch if you can clearly demonstrate that it doesn't cause any performance regressions.

Please refer to https://github.com/mozilla/pdf.js/wiki/Benchmarking-your-changes for information about benchmarking, and perhaps https://github.com/mozilla/pdf.js/wiki/Contributing is useful too for general information about the test-suite.

The following PDF files contains *lots* of rendering operations, and would probably be good examples to use when benchmarking:
https://github.com/mozilla/pdf.js/files/1522715/wuppertal_2012.pdf
https://www.prorail.nl/sites/default/files/spoorkaart_prorail_april_2013.pdf

**Comment by MeiKatz** [comment]
@Snuffleupagus Maybe I will turn this topic into another direction and write an example with improved readability while ignoring performance issues. So beginners would have a good point to start and can from there understand other implementation like the one for SVG or canvas.

**Comment by timvandermeij** [comment]
I think there are two problems here.

The most important one is that the increased number of function calls will have a bad effect on performance, and eliminating them is one of the reasons why #4683 worked so well.

Moreover, I don't think that the current code is hard to understand and I don't feel that this patch makes it any clearer. For instance, all function names are the same as the operator, so you get `this.rectangle(...)` just below `case OPS.rectangle` and that doesn't give any more context. The same goes for the parameters and variable names: previously it would be `x = args[j++];` whereas now it's `args[j++]; // x`, which is equivalent and also doesn't provide more context. The remaining comments are of the form `// ensure height is not null` just above `if (height === 0)`, which is redundant.

In conclusion, I'm closing this since there does not appear to be much benefit in merging this unfortunately. We can definitely consider a patch though that adds inline documentation as comments, but then they do have to add context and not just explain the code.

**Comment by MeiKatz** [comment]
I solved my readability issues by writing my own PDF.js core documentation that can be found [here](https://github.com/MeiKatz/pdfjs-docs/blob/master/README.md).

**Comment by timvandermeij** [comment]
Thank you for your contribution! I think it can definitely help in understanding the core of PDF.js and I think it's also much better suited on a specialized wiki page than in inline comments. I have linked the documentation on the https://github.com/mozilla/pdf.js/wiki/Additional-Learning-Resources page.

**Comment by MeiKatz** [comment]
> _Edit: Small feedback, it looks like the setTextRenderingIntent link in the overview does not work when clicked, while the others do._

I fixed it. The name `setTextRenderingIntent` was wrong because it is `setTextRenderingMode`.

I will try to improve my documentation when I have the time for it.
